core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,586 @@
+{-# LANGUAGE TypeOperators       #-}
+module Pos.Binary.Class.Core
+    ( Bi(..)
+    , encodeBinary
+    , decodeBinary
+    , enforceSize
+    , matchSize
+    -- * CBOR re-exports
+    , encodeListLen
+    , decodeListLen
+    , decodeListLenOf
+    , Encoding
+    , Decoder
+    -- * GHC-Generics-based encoding & decoding
+    , genericEncode
+    , genericDecode
+    ) where
 
-#include "MachDeps.h"
+import           Codec.CBOR.Decoding
+import           Codec.CBOR.Encoding
+import           GHC.Generics
+import qualified Data.Binary                 as Binary
+import qualified Data.ByteString             as BS
+import qualified Data.ByteString.Lazy        as BS.Lazy
+import           Data.Fixed                  (Fixed(..), Nano)
+import qualified Data.HashMap.Strict         as HM
+import qualified Data.HashSet                as HS
+import qualified Data.Map                    as M
+import qualified Data.Set                    as S
+import           Data.Tagged
+import qualified Data.Text                   as Text
+import           Data.Time.Units             (Microsecond, Millisecond)
+import qualified Data.Vector                 as Vector
+import qualified Data.Vector.Generic         as Vector.Generic
+import           Serokell.Data.Memory.Units  (Byte, fromBytes, toBytes)
+import           Universum                   hiding (foldr)
+import qualified Universum
 
--- | Bi typeclass and most basic functions.
+encodeBinary :: Binary.Binary a => a -> Encoding
+encodeBinary = encode . BS.Lazy.toStrict . Binary.encode
 
-module Pos.Binary.Class.Core
-       ( Bi (..)
-       , encode
-       , encodeWithS
-       , encodeLazy
-       , decodeFull
-       , decodeOrFail
-       , getSize
-       , biSize
-       , label
-       , labelP
-       , labelS
-       -- * Length things
-       , limitGet
-       , isolate64Full
-       , getPeekLength
-       ) where
-
-import           Universum
-
-import           Control.Lens               (_Left)
-import qualified Data.ByteString.Lazy       as BSL
-import           Data.Store                 (Size)
-import           Data.Store.Core            (Peek (..), PeekResult (..), Poke (..))
-import qualified Data.Store.Core            as Store
-import           Data.Store.Internal        (PeekException (..), PokeException (..))
-import qualified Data.Store.Internal        as Store
-import           Foreign.Ptr                (minusPtr, plusPtr)
-import           Formatting                 (build, sformat, shown, (%))
-import           Serokell.Data.Memory.Units (Byte)
+decodeBinary :: Binary.Binary a => Decoder s a
+decodeBinary = do
+    x <- decode @ByteString
+    case Binary.decodeOrFail (BS.Lazy.fromStrict x) of
+        Left (_, _, err) -> fail err
+        Right (bs, _, res)
+            | BS.Lazy.null bs -> pure res
+            | otherwise       -> fail "decodeBinary: unconsumed input"
+
+-- | Enforces that the input size is the same as the decoded one, failing in case it's not.
```

Comment:
> I don't understand what this function does. What size does it enforce? Size of a list that is ready to be consumed? (Only list, or does it work on bytestrings/etc too?)

We call `enforceSize` and `matchSize` when we expect a finite list length as input, and the specific case we cover is the general patter where each constructor with more than 1 argument needs to be encoded as a finite-length list, otherwise it wouldn't be a valid CBOR `FlatTerm` encoding.

> Also, another question: how will it work on indefinite-length lists?

It won't! This function should never be called with anything non-finite, as it has a very specific use case.